src/compile/codegen.rs

```diff
@@ -208,42 +202,71 @@ fn add_args_to_locals_table(
 /// statement sequence, otherwise the function returns a CodegenError.
 fn mavm_codegen_statements(
     statements: Vec<TypeCheckedStatement>, // statements to codegen
-    mut code: &mut Vec<Instruction>,       // accumulates the code as it's generated
+    code: &mut Vec<Instruction>,           // accumulates the code as it's generated
     mut num_locals: usize,                 // num locals that have been allocated
     locals: &CopyingSymTable<'_, usize>,   // lookup local variable slot number by name
     mut label_gen: LabelGenerator,
     string_table: &StringTable,
     import_func_map: &HashMap<StringId, Label>,
     global_var_map: &HashMap<StringId, usize>,
 ) -> Result<(LabelGenerator, usize, bool, HashMap<StringId, usize>), CodegenError> {
-    // (label_gen, num_labels, execution_might_continue)
-    if statements.is_empty() {
-        return Ok((label_gen, num_locals, true, HashMap::new()));
-    }
-    let rest_of_statements = &statements[1..];
-    match &statements[0] {
-        TypeCheckedStatement::Noop(_) => mavm_codegen_statements(
-            rest_of_statements.to_vec(),
+    let mut bindings = HashMap::new();
+    for statement in statements {
+        let new_locals = locals.push_multi(bindings.clone());
+        let (lg, statement_locals, returns, statement_bindings) = mavm_codegen_statement(
+            statement,
             code,
             num_locals,
-            locals,
+            &new_locals,
             label_gen,
             string_table,
             import_func_map,
             global_var_map,
-        ),
+        )?;
+        label_gen = lg;
+        num_locals = max(statement_locals, num_locals);
+        for (id, bind) in statement_bindings {
+            bindings.insert(id, bind);
+        }
+        if returns {
+            return Ok((label_gen, num_locals, false, bindings));
+        }
+    }
+    Ok((label_gen, num_locals, true, bindings))
+}
+
+///Generates code for the provided statement. code represents the code generated previously,
+/// num_locals the maximum number of locals used at any point in the call frame so far, locals is a
+/// map of local variables, label_gen points to the next available locals slot, string_table is used
+/// to get builtins, import_func_map associates each imported function with a label, and
+/// global_var_map maps global variable IDs to their slot number.
+///
+/// If successful the function returns a tuple containing the updated label generator, number of
+/// locals slots used by this statement, a bool that is set to true if the statement causes a
```

Comment:
It's hard to follow the logic of how locals are handled in this code.  (This has been true since the beginning, but this might be a good time to fix it.). Some statements, like `let`, allocate a new local.  Others, like codeblock statements, will allocate locals within their codegen and then those locals will go out of scope.  

I think it makes sense to refactor by creating a data structure that tracks locals allocation properly, rather than carrying around the separate `num_locals` and locals-map / binding objects and needing to maintain invariants about them everywhere.